### ISSUE TITLE
AvalonDock: Document doesn't show any content initially.

### ISSUE THREAD

**Issue by DoubleDHZ** [xceedsoftware/wpftoolkit#1597]
We want to bind our collection of ViewModels to the DocumentsSource. It works more or less, the only problem we have is, that when a new document is shown, it is empty. Only switching back and forth between existing tabs, shows the content correctly.
I've prepared a sample, which shows said behavior:

[AvalonDockSample.zip](https://github.com/xceedsoftware/wpftoolkit/files/4890198/AvalonDockSample.zip)

(A few words about the sample: The actual viewmodels (SampleViewModel in this case) shouldn't know anything about which 3rd party lib we use for the view, so we have this intermediate SynchronizingViewModelWrapper, which knows both our ViewModels and the LayoutDocuments and can mediate between them. That's why it might look a tad more complicated as necessary).

Anyways, if you start the sample and press add, you'll get an empty document. Add a second one, which also will be empty. Switch back to the first document and you'll see the actual content.

The question is, is this a bug? Or are we using AvalonDock wrong? If the latter is the case, how can we solve this issue?

**Comment by X39** [comment]
Can confirm it also happens in my project. Upgraded from 3.8.0 to 4.1 today.
Anchorables work fine, Documents do not.

**Comment by 1978tseng** [comment]
I am having the same issue 
but for my case , 
the first create content is not shown, 
it does not matter if it is  Anchorable or Document.

**Comment by XceedBoucherS** [comment]
Hi,
This issue will be fixed in v4.1 (latest version on github is v4.0.1).

In the meantime, you can go in file Xceed.Wpf.AvalonDock/Themes/generic.xaml,
Look for the 2 "PART_SelectedContentHost",
and remove their ContentTemplate binding.

**Comment by 1978tseng** [comment]
Thanks for the reply .
I have downgraded to 3.8
for some emergency build. 
and it is working at the moment.
I will try the File fix latter. 

Chun-Wei Tseng
David
810 282 1329 USA Michigan

**Comment by DoubleDHZ** [comment]
> In the meantime, you can go in file Xceed.Wpf.AvalonDock/Themes/generic.xaml,
> Look for the 2 "PART_SelectedContentHost",
> and remove their ContentTemplate binding.

Thanks for the heads-up!

It works much better with the workaround, but there are still some corner cases, where it stays empty. I can reproduce them every now and then, but haven't figured out, how to reproduce them reliably.

**Comment by XceedBoucherS** [comment]
Ok, if you can find a pattern to reproduce them, please tell us.
Thank you

**Comment by DoubleDHZ** [comment]
I can reliably reproduce it in the actual application, which I can't post here.
It is not possible to reproduce it in any sample application.

In the actual application we have two different ViewModels, we're displaying, let's call the A & B. Both of them derive from the same base class (in the sample app from above that base class would be SampleViewModel).
So ViewModel A works 100% reliable. It always displays correctly.
But ViewModel B has issues:

1. If no other documents are opened, It can open and close B without issue. Always displays correct.
2. If at least one document is opened (does not matter if A or B) and I open a B, it displays correctly. But if I close that newly created B and open yet another B, then the bug occurs and the tab stays empty. I can now open as much ViewModels of type B, they all stay empty. But I can still open ViewModels of type A and they still display correctly. On a faulted B nothing works, not closing, not any command from the context menu. I can switch from the faulted tab to another one and go back to the  faulted one, then I will see the content, just as described in my original post. But closing that tab or any context menu action still won't work.

What I can do is switching to a working tab and do a "close all tabs". This resets the behavior, I can open a working B again and both cases from above (1 & 2) apply again.

The best part is, that AvalonDock doesn't know anything about ViewModels A & B. AvalonDock just binds to a collection of SynchronizingViewModelWrappers, which wrap viewmodels A & B (just like it wraps SampleViewModel from the sample app). And there is also just one ILayoutUpdateStrategy, which handles the SynchronizingViewModelWrappers. Also just a single DataTemplate for SynchronizingViewModelWrappers. And you guessed it, only a single code path with creates the SynchronizingViewModelWrapper and adds it to the collection, to which the DockingManager binds.
It just doesn't make any sense at all!

I inspected the LayoutDocumentControl of a faulted tab in the visual tree, the only notable difference is, that the LayoutItem-Property has been set locally to null, as it appears:

![AvalonDock_LDC](https://user-images.githubusercontent.com/66508917/89188282-bcf5dd00-d59e-11ea-8b36-29bc97c9fbea.png)

Which as well doesn't make any sense at all, since it is a read-only dependency property.

I've no idea, if this helps you or not, but unfortunately this is as much insight as I can give.

**Comment by DoubleDHZ** [comment]
So users started to report issues, when opening tabs with ViewModel A as well. As this was only happening under certain circumstances, this narrowed down the possibilities enough, so I could finally create a new sample application to reliably replicate the issue:

[AvalonDockSample.zip](https://github.com/xceedsoftware/wpftoolkit/files/5027144/AvalonDockSample.zip)
(This one uses a custom AvalonDock build (Xceed.Wpf.AvalonDock.Hotfixed.dll), which includes the hotfix you suggested earlier in this thread, no other changes except the namespace was updated, so it doesn't conflict with the actual AvalonDock release)

How to reproduce:
Start the app & click "Add Item". Now you'll get a second "Add item" button, which is located within the tab. If you click that button, you'll always create a faulted tab, which initially doesn't show and content and no action works (not the "X" for closing nor any other action from the context menu).

The button not within a tab will always produce working tabs.
The button withing a tab will aways produce faulted tabs.
Both commands bound to each button use the same codepth to create & add the viewmodel to the collection. Nothing different here from our code's point of view.

**Comment by XceedBoucherS** [comment]
@DoubleDHZ , thank you for your clear explanation. 

I started your sample in a .NET 4.6.1 or .NET 4.6.2 environment and it worked as expected. The inside button can pop a valid tab.
I then started the same sample in a .NET 4.7.2 or 4.7.1 environment and the inside button popped a faulty tab.

Something must not work as expected in the sample (or AvalonDock) with latest .NET versions.

**Comment by mazurekj** [comment]
Yes. My company has the same issue. Our project needed to be upgraded from .NET 4.6.2 to .NET 4.7.2 and your DockingManager does not open new tabs correctly in some cases - they are blank. Before this update we used  Wpf Toolkit in 3.8.1. But upgrading your control to any higher version does not help. Can you please write, when will this problem be fixed?

**Comment by XceedBoucherS** [comment]
Hi,

A workaround for now would be to delay the "_layoutDocument.IsActive = true;" after the DataTemplate is applied for the "SynchronizingViewModelWrapper".
So in SynchronizingViewModelWrapper.Initialize(), you could do:
`Dispatcher.CurrentDispatcher.BeginInvoke( DispatcherPriority.SystemIdle, ( Action )delegate
          {
            _layoutDocument.IsActive = true;
          } );`
instead of : 
` _layoutDocument.IsActive = true;`
